lib/Sema/TypeCheckStmt.cpp

```diff
@@ -1737,10 +1737,74 @@ Stmt *StmtChecker::visitBraceStmt(BraceStmt *BS) {
   return BS;
 }
 
+static Optional<unsigned>
+getParamIndex(const ParameterList *paramList, const ParamDecl *decl) {
+  ArrayRef<ParamDecl *> params = paramList->getArray();
+  for (unsigned i = 0; i < params.size(); ++i) {
+    if (params[i] == decl) return i;
+  }
+  return None;
+}
+
+static void
+checkInheritedDefaultValueRestrictions(TypeChecker &TC, ParamDecl *PD) {
+  if (PD->getDefaultArgumentKind() != DefaultArgumentKind::Inherited)
+    return;
+
+  auto *DC = PD->getInnermostDeclContext();
+  const SourceFile *SF = DC->getParentSourceFile();
+  if (!SF || SF->Kind != SourceFileKind::Interface)
+    return;
+
+  if (!isa<ConstructorDecl>(DC)) {
+    TC.diagnose(PD->getLoc(),
+                diag::inherited_default_value_not_on_constructor_param);
+    return;
+  }
+
+  // Check the matching parameter of an overriden decl has a default value
+  auto ctor = cast<ConstructorDecl>(DC);
+  Optional<unsigned> idx = getParamIndex(ctor->getParameters(), PD);
+  bool foundOverriden = false;
+  assert(idx && "containing decl does not contain param?");
+  for (auto overridden = ctor->getOverriddenDecl();
+       overridden != nullptr;
+       overridden = overridden->getOverriddenDecl()) {
+    foundOverriden = true;
+    auto equivalentParam = overridden->getParameters()->get(*idx);
+    assert(equivalentParam && "inherited decl mismatched arguments?");
+
+    // If the corresponding overridden param is also inherited, keep looking
+    if (equivalentParam->getDefaultArgumentKind() ==
+        DefaultArgumentKind::Inherited)
+      continue;
+
+    // If it doesn't have a default value, diagnose it
+    if (equivalentParam->getDefaultArgumentKind() == DefaultArgumentKind::None) {
+      TC.diagnose(PD->getLoc(),
+                  diag::corresponding_param_not_defaulted);
+    }
+    // We found a match: it's valid
```

Comment:
…or we diagnosed, yeah?

Reply:
Oops, yeah. I reworked this section a bit, because I realized i didn't need the loop through the chain of overridden decls when the corresponding default is inherited too as that default be checked separately anyway.